### PR TITLE
[SYCL] Lift restrictions on free-function kernels when compiling at runtime

### PR DESCRIPTION
**Pull request by sommerlukas**
In order to be able to generate correct and complete information for the integration header, the current implementation places some restrictions on free-function kernels and their parameters. For example, parameters of free function kernels need to be forward-declarable. 

However, when compiling SYCL code at runtime (RTC), e.g., through the `kernel_compiler` extension, host code is typically not relevant, so the integration header is not as relevant and some restrictions on free-function kernels can be lifted. 

This PR introduces a `-fsycl-rtc-mode` flag (and it's negative equivalent) to deactivate some restrictions on free-function kernels and omit some information for free-function kernels from the integration header.

### CODE REVIEW

**Review by Naghasan — APPROVED** [review]
LGTM

**Review by elizabethandrews — COMMENTED** [review]
>>host code is typically not relevant, so the integration header is not as relevant and some restrictions on free-function kernels can be lifted.

Can you link the extension spec which explains this? What do you mean by `not as relevant`? Is it ever relevant in this mode? Should we be removing integration header functionality entirely in this mode?

**Comment by sommerlukas** [comment]
> > > host code is typically not relevant, so the integration header is not as relevant and some restrictions on free-function kernels can be lifted.
> 
> Can you link the extension spec which explains this? What do you mean by `not as relevant`? Is it ever relevant in this mode? 

@elizabethandrews Thanks for you feedback!

The extension specification for the `kernel_compiler` with SYCL as source language is in #11985. 

The extension defines an interface to compile kernels through the SYCL kernel bundle interface, but does not allow to compile additional host code. 

The current implementation is in [`kernel_compiler_sycl.cpp`](https://github.com/intel/llvm/blob/b7ef830fb490cf5b344035dfc0be7a78a9e6ccd5/sycl/source/detail/kernel_compiler/kernel_compiler_sycl.cpp). As `-fsycl-device-only`/`-c` and `-fsycl-dump-device-code` in combination do not produce finalized SPIR-V, the current implementation adds a dummy `main` for the host code and performs host and device compilation. The host code is however never used after the compilation, only the SPIR-V for the device is needed (see [line 138](https://github.com/intel/llvm/blob/b7ef830fb490cf5b344035dfc0be7a78a9e6ccd5/sycl/source/detail/kernel_compiler/kernel_compiler_sycl.cpp#L138),

> Should we be removing integration header functionality entirely in this mode?

This would also be an option, yes. For RTC, it would be useful to have a mode/flag to only compile device code, yielding finalized SPIR-V. No integration header would be needed in this mode and restrictions for free-function kernels would not need to be in place.

Removing the integration header entirely seemed quite invasive to me, that's why I chose this less-invasive route to address an immediate need for one of our projects. But I'm happy to have a conversation how we can improve things further.

File: clang/include/clang/Driver/Options.td [hunk 1]
```diff
@@ -6877,6 +6877,11 @@ defm sycl_esimd_force_stateless_mem : BoolFOption<"sycl-esimd-force-stateless-me
   NegFlag<SetFalse, [], [ClangOption, CLOption], "Do not enforce using "
           "stateless memory accesses.">,
   BothFlags<[], [ClangOption, CLOption, CC1Option], "">>;
+defm sycl_rtc_mode: BoolFOption<"sycl-rtc-mode",
+  LangOpts<"SYCLRTCMode">, DefaultFalse,
+  PosFlag<SetTrue, [], [ClangOption], "RTC Mode On">,
+  NegFlag<SetFalse, [], [ClangOption], "RTC Mode Off">,
```
Comment (mdtoguchi) [inline clang/include/clang/Driver/Options.td:6883]:
```suggestion
  PosFlag<SetTrue, [], [ClangOption], "Enable">,
  NegFlag<SetFalse, [], [ClangOption], "Disable">,
```
"RTC Mode On RTC mode in SYCL." didn't make much sense to me :)

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 2]
```diff
@@ -5583,6 +5583,9 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
       CmdArgs.push_back("-fsycl-allow-func-ptr");
     }
 
+    Args.AddLastArg(CmdArgs, options::OPT_fsycl_rtc_mode,
+                    options::OPT_fno_sycl_rtc_mode);
```
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/Clang.cpp:5587]: Can you add a small driver test to verify the passing of this to the expected `-cc1` phase?

Comment (sommerlukas) [reply clang/lib/Driver/ToolChains/Clang.cpp:5587]: Sure, no problem. I had looked around before posting the PR to see if other options handled similarly are tested in a certain manner, but didn't find a good example (probably me being blind). Do you happen to have a example of a good test for that?

File: clang/include/clang/Driver/Options.td (hunk 1 above)
Comment (sommerlukas) [reply clang/include/clang/Driver/Options.td]: Good catch, thanks!

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 2 above)
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/Clang.cpp:5587]: I think test/Driver/sycl-xarch.cpp has a good example.  The first test in the file has a check for the device compilation confirming the needed option is passed in that command.

File: clang/include/clang/Driver/Options.td [hunk 3]
```diff
@@ -6877,6 +6877,11 @@ defm sycl_esimd_force_stateless_mem : BoolFOption<"sycl-esimd-force-stateless-me
   NegFlag<SetFalse, [], [ClangOption, CLOption], "Do not enforce using "
           "stateless memory accesses.">,
   BothFlags<[], [ClangOption, CLOption, CC1Option], "">>;
+defm sycl_rtc_mode: BoolFOption<"sycl-rtc-mode",
+  LangOpts<"SYCLRTCMode">, DefaultFalse,
+  PosFlag<SetTrue, [], [ClangOption], "Enable">,
```
Comment (srividya-sundaram) [inline clang/include/clang/Driver/Options.td:6882]:
```suggestion
  PosFlag<SetTrue, [], [ClangOption], "Enable runtime compilation mode in SYCL">,
```

File: clang/include/clang/Driver/Options.td [hunk 4]
```diff
@@ -6877,6 +6877,11 @@ defm sycl_esimd_force_stateless_mem : BoolFOption<"sycl-esimd-force-stateless-me
...
+  NegFlag<SetFalse, [], [ClangOption], "Disable">,
```
Comment (srividya-sundaram) [inline clang/include/clang/Driver/Options.td:6883]:
```suggestion
  NegFlag<SetFalse, [], [ClangOption], "Disable runtime compilation mode in SYCL">,
```

File: clang/include/clang/Driver/Options.td (hunk 3 above)
Comment (sommerlukas) [reply clang/include/clang/Driver/Options.td:6882]: @srividya-sundaram Thanks for your feedback! After the improvements proposed by @mdtoguchi [here](https://github.com/intel/llvm/pull/15892#discussion_r1822878382), the help text already reads as follows:
```
 -fno-sycl-rtc-mode      Disable RTC mode in SYCL.
 -fsycl-rtc-mode         Enable RTC mode in SYCL.
```

Are you OK if we keep this as is?

**Comment by elizabethandrews** [comment]
> > Should we be removing integration header functionality entirely in this mode?
> 
> This would also be an option, yes. For RTC, it would be useful to have a mode/flag to only compile device code, yielding finalized SPIR-V. No integration header would be needed in this mode and restrictions for free-function kernels would not need to be in place.

I assume you mean restrictions added to free-function kernels - due to the integration header requirement - will no longer be needed. Other restrictions will still be required. 

> 
> Removing the integration header entirely seemed quite invasive to me, that's why I chose this less-invasive route to address an immediate need for one of our projects. But I'm happy to have a conversation how we can improve things further.

I don't think it is invasive if it is not required. If RTC does not require integration header, I would lean towards removing the functionality for this mode (not necessarily in this PR) but I would like @premanandrao and @tahonermann  to weigh in here.

**Comment by sommerlukas** [comment]
> > > Should we be removing integration header functionality entirely in this mode?
> > 
> > 
> > This would also be an option, yes. For RTC, it would be useful to have a mode/flag to only compile device code, yielding finalized SPIR-V. No integration header would be needed in this mode and restrictions for free-function kernels would not need to be in place.
> 
> I assume you mean restrictions added to free-function kernels - due to the integration header requirement - will no longer be needed. Other restrictions will still be required.

Yes, this PR was specifically for restrictions stemming from the integration header.

> > Removing the integration header entirely seemed quite invasive to me, that's why I chose this less-invasive route to address an immediate need for one of our projects. But I'm happy to have a conversation how we can improve things further.
> 
> I don't think it is invasive if it is not required. If RTC does not require integration header, I would lean towards removing the functionality for this mode (not necessarily in this PR) but I would like @premanandrao and @tahonermann to weigh in here.

IIUC, there is also some ongoing/planned effort to replace the integration header with another mechanism. Potentially, we could land the change in this PR as a workaround for now and remove the hidden flag again once the integration header mechanism has been replaced.

**Comment by tahonermann** [comment]
> IIUC, there is also some ongoing/planned effort to replace the integration header with another mechanism.

That mechanism will only work with SYCL-aware host compilers. The integration header approach will still be needed to support third party host compilers (if such support is required).

File: clang/include/clang/Driver/Options.td (hunk 3 above)
Comment (srividya-sundaram) [reply clang/include/clang/Driver/Options.td:6882]: Thanks for the clarification. I just realized that the text in `Pos/Neg` flags get appended with the text in `BothFlags`.
I was wondering why they just read "Enable" and "Disable" only :)
I am OK with the texts here.

**Comment by gmlueck** [comment]
It seems like this option is only being added as a short-term hack because the kernel compiler implementation currently invokes the host compilation pass even though there is no need to generate host code.  Once that is fixed, I think we will no longer need this option.

Given that it is just a short-term hack, is this the right way to define the option?  Should we instead pass the option via `-Xclang`?  Would that make it easier for us to remove the option later?

**Comment by elizabethandrews** [comment]
> It seems like this option is only being added as a short-term hack because the kernel compiler implementation currently invokes the host compilation pass even though there is no need to generate host code. Once that is fixed, I think we will no longer need this option.

Integration header generation happens during device compilation. So these diagnostics will fire even if host compilation is disabled. So I do not think this option is temporary unless we come up with another way to tell the compiler that we do not need to generate the integration header. 

> 
> Given that it is just a short-term hack, is this the right way to define the option? Should we instead pass the option via `-Xclang`? Would that make it easier for us to remove the option later?

From an implementation perspective I don't think it matters.

**Comment by gmlueck** [comment]
> Integration header generation happens during device compilation. So these diagnostics will fire even if host compilation is disabled. So I do not think this option is temporary unless we come up with another way to tell the compiler that we do not need to generate the integration header.

I'm pretty sure we will not expose the option to the user with this name.  For example, if we use "-fsyclbin" to compile SYCLBIN files as I proposed on Wednesday, we would tie this feature to the "-fsyclbin" option.  However, even that option name is not decided.

> From an implementation perspective I don't think it matters.

I meant more from an API stability point of view.  If we document "-fsycl-rtc-mode" are we committing to support it forever?  I think we want to avoid any commitment to support this option.

**Comment by sommerlukas** [comment]
> > From an implementation perspective I don't think it matters.
> 
> I meant more from an API stability point of view. If we document "-fsycl-rtc-mode" are we committing to support it forever? I think we want to avoid any commitment to support this option.

The option is hidden, i.e., it does not appear in `clang++ --help` and I intentionally did not add it to the user manual, so we don't really expose it ATM. IMHO, this would not imply a commitment to support this option in the future.

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 2 above)
Comment (sommerlukas) [reply clang/lib/Driver/ToolChains/Clang.cpp:5587]: Thanks for the hint, I've added a test now.

**Comment by elizabethandrews** [comment]
> I'm pretty sure we will not expose the option to the user with this name. For example, if we use "-fsyclbin" to compile SYCLBIN files as I proposed on Wednesday, we would tie this feature to the "-fsyclbin" option. However, even that option name is not decided.

I understand. My point was that conditionalizing the diagnostics (irrespective of what the option is called) is not temporary. This is a permanent change for this mode. And while it is ok as a quick workaround, eventually we need to decouple the generation of the integration from this mode.

**Review by elizabethandrews — APPROVED** [review]
I'm approving this for now but please file an issue to remove remove the generation of Integration header in this mode.

**Review by mdtoguchi — APPROVED** [review]

**Comment by sommerlukas** [comment]
> I'm approving this for now but please file an issue to remove remove the generation of Integration header in this mode.

Thanks, I've filed #15996.

**Pull request merged**